R/super.R

```diff
@@ -36,20 +36,80 @@
 super <- function(...) {
   # TODO: (RK) Error on anonymous function calls.
   fn <- as.character(sys.call(-1)[[1]])
-
-  env <- parent.frame(2)
+  if (identical(fn[1], "get")) {
+    if (length(fn) == 1) { # TODO: (RK) Actually parse out super::multi_parent_env
+      stop("super::super does not support ", sQuote("get"), " calls ",
+           "due to internal purposes.")
+    } else {
+      # From the magic get() call below, we extract the function name.
+      fn <- fn[2] 
+      env <- parent.frame()
+      level <- 2 # See comment below.
+    }
+  } else {
+    env   <- parent.frame(2)
+    level <- 7
+  }
 
   first <- TRUE
+  parent_count <- -1
   while (!identical(env, emptyenv())) {
+    parent_count <- parent_count + 1
     if (exists(fn, envir = env, inherits = FALSE)) {
       super_fn <- get(fn, envir = env, inherits = FALSE)
       if (is.function(super_fn)) {
         if (first) { first <- FALSE }
-        else { return(do.call(fn, eval(substitute(alist(...))), envir = env)) }
+        else {
+          # We have to find the equivalent of fn in the environment that is the
+          # n-fold parent environment of the calling environment, parent.frame(),
+          # where n = parent_count. Otherwise, non-standard evaluation will not work
+          # correctly with super.
+          return(eval.parent(substitute(
+            # The magic number level = 7 comes from the fact eval.parent(...) creates its own
+            # call chain with 5 steps inserted that we wish to skip over. Otherwise,
+            # if we are already in a super calling chain, we use level = 2.
```

Comment:
In English: in the calling environment, take the last system call, replace its caller (`[[1]]`) with the next rightful `fn` as already computed using the `hop`s above (note that `super::multi_parent_env(env, n)` is simply `n` `hop`s from `env`), and execute that call (again, in the calling environment!) using **exactly the same arguments**.

This approach keeps promises intact and allows non-standard evaluation in the calling chain.

To really go full steam here, we would have to do something extraordinarily clever: we would have to call `match.call` on `sys.call(-4)`, substitute those arguments directly into the `super` call (`sys.call(-1)`) using the named expressions, and then evaluate this in the parent frame.